src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+                }
+                jGraphic.element(FORMAT, em.getFormat());
+            }
+        } else if (g instanceof ExternalGraphic) {
+            ExternalGraphic eg = (ExternalGraphic) g;
+            try {
+                jGraphic.element(
+                        EXTERNAL_GRAPHIC_URL,
+                        eg.getOnlineResource().getLinkage().toURL().toString());
+            } catch (MalformedURLException e) {
+                e.printStackTrace();
+            }
+            jGraphic.element(EXTERNAL_GRAPHIC_TYPE, eg.getFormat());
+        }
+        return jGraphic;
+    }
+
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processLineSymbolizer(JSONObject ret, LineSymbolizer symbolizer) {
+        ret = processStroke(ret, symbolizer.getStroke());
+        if (symbolizer.getPerpendicularOffset() != null) {
+            ret.element(
+                    PERPENDICULAR_OFFSET,
+                    printExpression(symbolizer.getPerpendicularOffset(), Number.class));
+        }
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processPointSymbolizer(JSONObject ret, PointSymbolizer symbolizer) {
+        ret = processGraphic(ret, symbolizer.getGraphic());
+        return ret;
+    }
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processPolygonSymbolizer(JSONObject ret, PolygonSymbolizer symbolizer) {
+        ret = processStroke(ret, symbolizer.getStroke());
+        ret = processFill(ret, symbolizer.getFill());
+        return ret;
+    }
+    /**
+     * @param ret
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processRasterSymbolizer(JSONObject ret, RasterSymbolizer symbolizer) {
+        ret = processColorMap(symbolizer.getColorMap(), ret);
+        Expression op = symbolizer.getOpacity();
+        if (op != null) {
+            ret.element(OPACITY, printExpression(op, Double.class));
+        }
+        ret = processChannelSelection(ret, symbolizer.getChannelSelection());
+        ret = processContrastEnhancement(ret, symbolizer.getContrastEnhancement());
+        return ret;
+    }
+    /**
+     * @param ret
+     * @param contrastEnhancement
+     * @return
+     */
+    private JSONObject processContrastEnhancement(
+            JSONObject ret, ContrastEnhancement contrastEnhancement) {
+        if (contrastEnhancement != null) {
+            JSONObject ce = new JSONObject();
+            Expression gammaValue = contrastEnhancement.getGammaValue();
+            if (gammaValue != null)
+                ce.element(GAMMA_VALUE, printExpression(gammaValue, Number.class));
+            ContrastMethod method = contrastEnhancement.getMethod();
+            if (ContrastMethod.NORMALIZE == method) {
+                JSONObject norm = new JSONObject();
+                // TODO handle vendor option
```

Comment:
Leftover TODO, if does not take too much time, could be useful to see the vendor options too

Reply:
done - though annoyingly broke my vendor option handler!